nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/nifi/processors/hive/UpdateHive3Table.java

```diff
@@ -0,0 +1,470 @@
+            Statement s = conn.createStatement();
+            ResultSet tables = s.executeQuery("SHOW TABLES");
+            List<String> tableNames = new ArrayList<>();
+            String hiveTableName;
+            while (tables.next() && StringUtils.isNotEmpty(hiveTableName = tables.getString(1))) {
+                tableNames.add(hiveTableName);
+            }
+
+            List<String> columnsToAdd = new ArrayList<>();
+            String outputPath;
+            if (!tableNames.contains(tableName) && createIfNotExists) {
+                StringBuilder createTableStatement = new StringBuilder();
+                for (RecordField recordField : schema.getFields()) {
+                    String recordFieldName = recordField.getFieldName();
+                    // The field does not exist in the table, add it
+                    columnsToAdd.add(recordFieldName + " " + NiFiOrcUtils.getHiveTypeFromFieldType(recordField.getDataType(), true));
+                    getLogger().debug("Adding column " + recordFieldName + " to table " + tableName);
+                }
+                createTableStatement.append("CREATE TABLE IF NOT EXISTS ")
+                        .append(tableName)
+                        .append(" (")
+                        .append(String.join(", ", columnsToAdd))
+                        .append(") STORED AS ")
+                        .append(storageFormat);
+
+                String createTableSql = createTableStatement.toString();
+
+                if (StringUtils.isNotEmpty(createTableSql)) {
+                    // Perform the table create
+                    getLogger().info("Executing Hive DDL: " + createTableSql);
+                    s.execute(createTableSql);
+                }
+
+                // Now that the table is created, describe it and determine its location (for placing the flowfile downstream)
+                String describeTable = "DESC FORMATTED " + tableName;
+                ResultSet tableInfo = s.executeQuery(describeTable);
+                boolean moreRows = tableInfo.next();
+                boolean locationFound = false;
+                while (moreRows && !locationFound) {
+                    String line = tableInfo.getString(1);
+                    if (line.startsWith("Location:")) {
+                        locationFound = true;
+                        continue; // Don't do a next() here, need to get the second column value
+                    }
+                    moreRows = tableInfo.next();
+                }
+                outputPath = tableInfo.getString(2);
+
+            } else {
+                List<String> hiveColumns = new ArrayList<>();
+
+                String describeTable = "DESC FORMATTED " + tableName;
+                ResultSet tableInfo = s.executeQuery(describeTable);
+                // Result is 3 columns, col_name, data_type, comment. Skip the first row as that just contains the header with column names
+                tableInfo.next();
```

Comment:
@mattyb149 I'm testing it on Hive 3.1.x and the first row contains the metadata of the first column of the table so it should not be skipped. It leads to an error later when the processor tries to re-add the existing column.

Reply:
Good catch, looks like some versions of Hive output a header row and others don't, I'll make it optional.